src/pytorch_lightning/trainer/trainer.py

```diff
@@ -654,12 +654,16 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self._call_callback_hooks("on_exception", exception)
+                for logger in self.loggers:
+                    logger.finalize("failed")
         except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaining processes, kill otherwise
                 self.strategy.reconciliate_processes(traceback.format_exc())
             self._call_callback_hooks("on_exception", exception)
+            for logger in self.loggers:
+                logger.finalize("failed")
```

Comment:
Note that with strategies like ddp_spawn, this won't do anything. On the main process, the logger has never done anything, so sending a "failed" signal will be wrong. 

There is currently no way we can handle this situation, and this is mostly the reason why only the "success" calls exist today. Sorry for the bad news :(

Reply:
Thank you for sharing. Is there anything I can do to help with the situation? I don't know If I have got the skills to do that.

Reply:
A quick temporary fix is to do this in mlflowlogger.finalize:

```py
if self._mlflow_client is None:
    return

# else do your thing here to set the status to failed
```

This logic relies on the fact that _mlflow_client is only ever not None in the worker process. In the main process it is None unless the user called the logger manually. 

This "hack" is not ideal, because it shifts responsibility to the logger that should ideally be handled by probably the launcher (i.e., only calling the finalize on the right process). 
This mechanism to process exceptions before they bubble up to the main process would be necessary to have for handling this properly. It needs some brainstorming, so probably the next step is to open a GH issue. 

For now, let's try to go with my suggestion and try to land the PR, what do you think? 

Note the above snippet would have to be adapted for each logger that implements finalize, i.e., checking whether `self._experiment is None`.

Reply:
@awaelchli 
> checking whether self._experiment is None.

In addition to the fact that objects to be checked were different for each logger, it is unclear how other loggers behave when `finalize("failed")` is called.
So I fixed it so that `finalize("failed")` is called on an exception only when the logger is mlfow. Can you review the new commits?
A quick overview of the changes is as follows:

trainer.py:
```py
def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs: Any) -> Any:
        ...
        except BaseException as exception:
            ...
            for logger in self.loggers:
                logger.finalize_when_exception()
```

logger.py:
```py
class Logger(ABC):
    ...
    def finalize_when_exception(self) -> None:
        pass
```

mlflow.py: 
```py
class MLFlowLogger(Logger):
    ...
    def finalize_when_exception(self) -> None:
        super().finalize_when_exception()
        if self._mlflow_client is None:
            return
        self.finalize("failed")
```

Reply:
> In addition to the fact that objects to be checked were different for each logger, it is unclear how other loggers behave when finalize("failed") is called.

If they don't do anything now, then we leave it like that. 

I wouldn't introduce any new api here, especially not in the base class when it's just affecting one logger. Did my [suggestion](https://github.com/Lightning-AI/lightning/pull/12292#discussion_r961550279) not work?
Happy to help and get hands-on here. Should I push a commit? I can take care of the other loggers.

Reply:
@awaelchli It would be helpful if you push a commit or give me some advice!

> Did my https://github.com/Lightning-AI/lightning/pull/12292#discussion_r961550279 not work?

If you fix `tensorboard.py` like this: 

```py
    @rank_zero_only
    def finalize(self, status: str) -> None:
        if self._experiment is None:
            return
        self.experiment.flush()
        self.experiment.close()
        self.save()
```

Then `tests/tests_pytorch/loggers/test_tensorboard.py::test_tensorboard_save_hparams_to_yaml_once` fails: 

```py
    def test_tensorboard_save_hparams_to_yaml_once(tmpdir):
        model = BoringModel()
        logger = TensorBoardLogger(save_dir=tmpdir, default_hp_metric=False)
        trainer = Trainer(max_steps=1, default_root_dir=tmpdir, logger=logger)
        assert trainer.log_dir == trainer.logger.log_dir
        trainer.fit(model)
    
        hparams_file = "hparams.yaml"
>       assert os.path.isfile(os.path.join(trainer.log_dir, hparams_file))
E       AssertionError: assert False
E        +  where False = <function isfile at 0x1059b8670>('/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0/hparams.yaml')
E        +    where <function isfile at 0x1059b8670> = <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'>.isfile
E        +      where <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'> = os.path
E        +    and   '/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0/hparams.yaml' = <function join at 0x1059b8e50>('/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0', 'hparams.yaml')
E        +      where <function join at 0x1059b8e50> = <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'>.join
E        +        where <module 'posixpath' from '/Users/ritsuki.yamada/.pyenv/versions/3.9.13/lib/python3.9/posixpath.py'> = os.path
E        +      and   '/private/var/folders/01/y61dt9vx52lds6ylr3975mbc0000gq/T/pytest-of-ritsuki.yamada/pytest-93/test_tensorboard_save_hparams_0/lightning_logs/version_0' = <pytorch_lightning.trainer.trainer.Trainer object at 0x130584dc0>.log_dir

tests/tests_pytorch/loggers/test_tensorboard.py:291: AssertionError
```

Reply:
@ritsuki1227 I noticed this too. This was because the save() in finalize of TensorBoard was expected to be called, even when no self.experiment is initialized. This is because the saving of hparams happens independently of the tensorboard summary writer. Now, I implemented it so that it behaves the same as on master. 

Otherwise, I took care of:
- Other loggers
- Improving the tests

I let the CI run now and check back later. If other tests fail, I will take another look.